### PR TITLE
593 mapper performance

### LINKED ISSUE #593: Optimize mapper performance
Use HashMap instead of iteration where possible (example: `get_operation` at https://github.com/qoollo/bob/blob/master/bob-common/src/mapper.rs#L184).
Avoid data cloning where possible (example: `DiskPath::from` at https://github.com/qoollo/bob/blob/master/bob-common/src/data.rs#L283)

### PR DESCRIPTION
**Pull request by lostl1ght**
Closes #593

### CODE REVIEW

**Review by idruzhitskiy — APPROVED** [review]

File: bob-common/src/data.rs [hunk 1]
```diff
@@ -260,14 +261,17 @@ impl VDisk {
     }
 
     pub fn push_replica(&mut self, value: NodeDisk) {
-        self.replicas.push(value)
+        self.replicas
+            .entry(value.node_name().to_string())
```
Comment (ikopylov) [inline bob-common/src/data.rs:265]: Why `to_string()` is called? It is `Arc<String>`

File: bob-common/src/data.rs [hunk 2]
```diff
@@ -260,14 +261,17 @@ impl VDisk {
...
+            .and_modify(|val| val.push(value.clone()))
+            .or_insert_with(|| vec![value]);
     }
 
     pub fn set_nodes(&mut self, nodes: &NodesMap) {
         nodes.values().for_each(|node| {
-            if self.replicas.iter().any(|r| r.node_name() == node.name()) {
+            if self.replicas.get(node.name()).is_some() {
```
Comment (ikopylov) [inline bob-common/src/data.rs:272]: `contains_key`

File: bob-common/src/data.rs [hunk 3]
```diff
@@ -260,14 +261,17 @@ impl VDisk {
...
                 //TODO check if some duplicates
```
Comment (ikopylov) [inline bob-common/src/data.rs:273]: Fix this

File: bob-common/src/mapper.rs [hunk 4]
```diff
@@ -181,13 +180,14 @@ impl Virtual {
...
         let virt_disk = self.get_vdisk_for_key(key).expect("vdisk not found");
-        let disk = virt_disk.replicas().iter().find_map(|disk| {
-            if disk.node_name() == self.local_node_name {
-                Some(DiskPath::from(disk))
-            } else {
-                None
-            }
-        }); //TODO prepare at start?
+        debug!("replicas: {:?}", virt_disk.replicas());
+        debug!("local_node: {:?}", self.local_node_name);
```
Comment (ikopylov) [inline bob-common/src/mapper.rs:184]: Do not split logs. This should  be a single call to `debug!`

File: bob-common/src/mapper.rs [hunk 5]
```diff
@@ -181,13 +180,14 @@ impl Virtual {
...
+        let disk = virt_disk
+            .replicas()
+            .get(&self.local_node_name)
+            .and_then(|replicas| {
+                replicas.get(0).and_then(|replica| Some(DiskPath::from(replica)))
```
Comment (ikopylov) [inline bob-common/src/mapper.rs:189]: I think the idea of `and_then` is to call them in a chain: `.and_then(...).and_then(...)`

File: bob/src/api/mod.rs [hunk 6]
```diff
@@ -270,15 +270,18 @@ fn find_vdisk<A: Authenticator>(bob: &BobServer<A>, id: u32) -> Option<&DataVDis
...
-            disk: r.disk_name().to_owned(),
-            node: r.node_name().to_owned(),
-        })
-        .collect()
+fn collect_replicas_info(replicas: &HashMap<String, Vec<NodeDisk>>) -> Vec<Replica> {
+    let mut vec_replicas = Vec::new();
+    for (_, nodes) in replicas {
+        for disk in nodes {
+            vec_replicas.push(Replica {
+                path: disk.disk_name().to_string(),
```
Comment (ikopylov) [inline bob/src/api/mod.rs:278]:
```suggestion
                path: disk.disk_path().to_string(),
```

File: bob-common/src/mapper.rs (hunk 4 above)
Comment (ikopylov) [reply bob-common/src/mapper.rs:184]: Also, I would prefer to write this on a `trace!` level

File: bob-common/src/data.rs [hunk 7]
```diff
@@ -234,15 +235,15 @@ impl BobOptions {
 #[derive(Debug, Clone)]
 pub struct VDisk {
     id: VDiskId,
-    replicas: Vec<NodeDisk>,
+    replicas: HashMap<String, Vec<NodeDisk>>,
```
Comment (ikopylov) [inline bob-common/src/data.rs:238]: I ran simple performance tests. Unfortunately, `HashMap` is slower than a vector when the number of elements is small (HashMap becomes faster with 10 elements or more in my tests). The usual number of replicas for Bob is 2 or 3. So this only reduces performance.
Unfortunately, this change must be reverted. Also please add a comment explaining why `Vec` is faster here

**Comment by ikopylov** [comment]
Will be replaced by https://github.com/qoollo/bob/pull/775

**Comment by ikopylov** [comment]
Replaced by https://github.com/qoollo/bob/pull/775

**Pull request closed without merging**